### ISSUE TITLE
Investigate sensor fusion

### ISSUE THREAD

**Issue by bgottula** [seeing-things/track#120]
The current hybrid error source switches between the blind and optical error sources such that only one of these is in use at a time and the other is completely ignored. This approach is simple but there may be better approaches that can continue to make use of the blind error source when the optical source is available. In general, combining of sensory data is known as _sensor fusion._ Investigate this subject area and identify which algorithms might be applicable.

Some algorithms that may apply:

- Complementary Filter https://www.youtube.com/watch?v=whSw42XddsU  (group delay of the extra filters required for this approach is concerning; may cause instability)
- Extended Kalman Filter (this seems the most promising; unclear if this also has group delay concerns)
  - To handle multiple sensors, can preform the "update" step twice, once for encoder sensors and once for camera sensor. Can use different observation noise covariance matrices for each sensor. See https://stackoverflow.com/a/55819191/2475856. Should also strongly consider alternating between blind and optical error sources on each control cycle to keep the cycle fast; computing both sets of error terms in a single cycle may be too slow.
  - The state vector can contain entries such as bias terms that are estimated but are not directly used or observed. In this case, the set of mount model parameters might be appropriate. If done correctly, this should allow the filter to estimate and remove biases in the blind error source on the fly such that it agrees more closely with the optical error source.

**Comment by bgottula** [comment]
I performed the first test of sensor fusion this evening by pointing at stars and Mars and after a fixing a few minor coding errors it worked on the first try! Here's what I learned:

- Works on both sides of meridian
- Integrator gain:
  - 1e-3 too slow
  - 1e-2 is slow but may be about right; need to test with satellites to know for certain
  - 1e-1 leads to ringing before settling down
- slew command lateness varies between 0 and 60 ms, which seems pretty high compared to other scenarios. Probably means more optimization is needed or a longer control cycle or only update the camera every other control cycle or something like that.

**Comment by bgottula** [comment]
I should note that I did some derivations on the whiteboard a while back that showed that the Kalman filter for estimating the bias in the blind target prediction using the camera was basically equivalent to a leaky integrator on the error between the two predicted positions. This derivation assumes that the system is running open loop--in other words, the estimated bias is not fed back and used to correct the pointing. So for the actual implementation I replaced the leaky integrator with a true integrator with a configurable gain parameter. This seems to work pretty well, despite perhaps not being as rigorously derived as I was hoping for.
